packages/components/src/components/viewers/JsonView.tsx

```diff
@@ -0,0 +1,78 @@
+import { useEffect, useState } from 'react'
+import type { FileSource } from '../../lib/sources/types.js'
+import { parseFileSize } from '../../lib/utils.js'
+import Json from '../Json.js'
+import { Spinner } from '../Layout.js'
+import ContentHeader, { TextContent } from './ContentHeader.js'
+import styles from '../../styles/Json.module.css'
+
+interface ViewerProps {
+  source: FileSource
+  setError: (error: Error | undefined) => void
+}
+
+const largeFileSize = 8_000_000 // 8 mb
+
+/**
+ * JSON viewer component.
+ */
+export default function JsonView({ source, setError }: ViewerProps) {
+  const [content, setContent] = useState<TextContent>()
+  const [json, setJson] = useState<unknown>()
+  const [isLoading, setIsLoading] = useState(true)
+
+  const { resolveUrl, requestInit } = source
+
+  // Load json content
+  useEffect(() => {
+    async function loadContent() {
+      try {
+        const res = await fetch(resolveUrl, requestInit)
+        const text = await res.text()
+        const fileSize = parseFileSize(res.headers) ?? text.length
```

Comment:
we might want to parse the headers and return an error earlier (before awaiting res.text()) if the file size is provided and above the threshold.

Another possible improvement, if it's not possible and the size is unknown before fetching: fetch with a range request (0 - 1M) and check the `content-length` and `content-range` headers of the response to see if the file is too big: /https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests#requesting_a_specific_range_from_a_server

Reply:
Refactored to return faster if content-length header is present and is too long. Good call!